test/FullRange.t.sol

```diff
@@ -0,0 +1,873 @@
+import {PoolModifyPositionTest} from "@uniswap/v4-core/contracts/test/PoolModifyPositionTest.sol";
+import {PoolSwapTest} from "@uniswap/v4-core/contracts/test/PoolSwapTest.sol";
+import {TickMath} from "@uniswap/v4-core/contracts/libraries/TickMath.sol";
+import {FullMath} from "@uniswap/v4-core/contracts/libraries/FullMath.sol";
+import {Oracle} from "../contracts/libraries/Oracle.sol";
+import {IHooks} from "@uniswap/v4-core/contracts/interfaces/IHooks.sol";
+import {UniswapV4ERC20} from "../contracts/hooks/UniswapV4ERC20.sol";
+import "@uniswap/v4-core/contracts/libraries/FixedPoint128.sol";
+
+import "forge-std/console.sol";
+
+contract TestFullRange is Test, Deployers, GasSnapshot {
+    using PoolIdLibrary for IPoolManager.PoolKey;
+
+    event Initialize(
+        PoolId indexed poolId,
+        Currency indexed currency0,
+        Currency indexed currency1,
+        uint24 fee,
+        int24 tickSpacing,
+        IHooks hooks
+    );
+    event ModifyPosition(
+        PoolId indexed poolId, address indexed sender, int24 tickLower, int24 tickUpper, int256 liquidityDelta
+    );
+    event Swap(
+        PoolId indexed id,
+        address indexed sender,
+        int128 amount0,
+        int128 amount1,
+        uint160 sqrtPriceX96,
+        uint128 liquidity,
+        int24 tick,
+        uint24 fee
+    );
+
+    int24 constant TICK_SPACING = 60;
+    uint160 constant SQRT_RATIO_2_1 = 112045541949572279837463876454;
+
+    /// @dev Min tick for full range with tick spacing of 60
+    int24 internal constant MIN_TICK = -887220;
+    /// @dev Max tick for full range with tick spacing of 60
+    int24 internal constant MAX_TICK = -MIN_TICK;
+
+    TestERC20 token0;
+    TestERC20 token1;
+    PoolManager manager;
+    FullRangeImplementation fullRange = FullRangeImplementation(
+        address(uint160(Hooks.BEFORE_INITIALIZE_FLAG | Hooks.BEFORE_MODIFY_POSITION_FLAG | Hooks.BEFORE_SWAP_FLAG))
+    );
+    IPoolManager.PoolKey key;
+    PoolId id;
+
+    // the key that includes a pool fee for pool fee rebalance tests
+    IPoolManager.PoolKey feeKey;
+    PoolId feeId;
+
+    PoolModifyPositionTest modifyPositionRouter;
+    PoolSwapTest swapRouter;
+
+    function setUp() public {
+        token0 = new TestERC20(2**128);
+        token1 = new TestERC20(2**128);
+        manager = new PoolManager(500000);
+
+        vm.record();
+        FullRangeImplementation impl = new FullRangeImplementation(manager, fullRange);
+        (, bytes32[] memory writes) = vm.accesses(address(impl));
+        vm.etch(address(fullRange), address(impl).code);
+        // for each storage key that was written during the hook implementation, copy the value over
+        unchecked {
+            for (uint256 i = 0; i < writes.length; i++) {
+                bytes32 slot = writes[i];
+                vm.store(address(fullRange), slot, vm.load(address(impl), slot));
+            }
+        }
+        key = IPoolManager.PoolKey(
```

Comment:
I think I had mentioned maybe starting a test with no fee for easiness, but I'm not sure this merits so much testing?? Feels like an edgecase